pkg/metadata.go

```diff
@@ -182,14 +216,33 @@ func (m Metadata) ApplyPolicy(p *MetadataPolicies) (*Metadata, error) {
 		}
 		reflect.Indirect(reflect.ValueOf(out)).Field(i).Set(reflect.ValueOf(applied))
 	}
-	return out, nil
-}
 
-type metadatas interface {
-	*OpenIDProviderMetadata | *OpenIDRelyingPartyMetadata | *OAuthAuthorizationServerMetadata | *OAuthClientMetadata | *OAuthProtectedResourceMetadata | *FederationEntityMetadata
+	// Iterate over extra metadata and associated policies
+	if len(m.Extra) > 0 {
+		out.Extra = map[string]interface{}{}
+		for entityType, metadata := range m.Extra {
+			var metadataToReturn interface{}
+			if policy, ok := p.Extra[entityType]; ok {
+				// Found a policy for the entity type, so apply it
+				applied, err := applyPolicy(metadata, policy, entityType)
+				if err != nil {
+					return nil, err
+				}
+
+				metadataToReturn = applied
+			} else {
+				// No policy found, so copy the metadata into out
+				metadataToReturn = metadata
+			}
+
+			out.Extra[entityType] = metadataToReturn
+		}
+	}
+
+	return out, nil
 }
 
-func applyPolicy[M metadatas](metadata M, policy MetadataPolicy, ownTag string) (any, error) {
+func applyPolicy(metadata any, policy MetadataPolicy, ownTag string) (any, error) {
```

Comment:
I removed the generic type parameter `M metadatas` here. The interface `metadatas` was just the union over the metadata types known to this Go module, but my goal is to allow inserting arbitrary extra `map[string]any` metadata dictionaries. And in any case, this function, which was the sole place where the generic parameter was used, didn't actually use anything about the `metadatas` interface. Regardless of the type of `metadata`, `applyPolicy` uses reflection to look for certain fields (e.g., `wasSet`) with certain types or tags, so we don't lose any safety by making the type of `metadata` `any`.